### ISSUE TITLE
[New rule] avoid_empty_nullable_iterables

### ISSUE THREAD

**Issue by kurtdnyc** [dart-code-checker/dart-code-metrics#1012]
**Please describe what the rule should do:**

It suggests you avoid passing in values to object constructors that are nullable, and iterable. If an object has a nullable, iterable property, assigning it an empty value allows an avoidable `StateError` to be thrown when accessing the `first` or `last` property.  Instead, it should be ensured that this property is either null, or not empty only. 

Additionally, it suggests you sanitize fields that aren't `final`, and within non-`const` constructors. 

This is especially common when creating objects from an HTTP response. Common practice for API's is to return an empty list, instead of omitting the field in a JSON response. 

**What category of rule is this? (place an "X" next to just one item)**

- [X] Warns about a potential error (problem)
- [ ] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

**❌ Bad:**

```dart
class SomeObject {
  List<String>? children; 
  SomeObject(this.children); // LINT
}

void main() async {
  SomeObject someObject = SomeObject([]); // LINT
  someObject.children?.first; // throws a StateError
}
```

**✅ Good:**

```dart
class SomeObject {
  List<String>? children;
  SomeObject(List<String>? children) {
    this.children = children?.isEmpty == true ? null : children;
  }
}

void main() async {
  SomeObject someObject = SomeObject(null);
  someObject.children?.first; // simply returns null
}
```

**❌ Bad:**

```dart
class SomeWidget extends StatelessWidget {
  const SomeWidget({Key? key}) : super(key: key);
  
  Future<List<String>?> _getStrings() async {
    //... 
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: _getStrings(),
      builder: (BuildContext context, AsyncSnapshot<List<String>?> snapshot) {
        if(snapshot.hasData){
          return OtherWidget(
            strings: snapshot.data, // LINT
          ); 
        }
        return const CircularProgressIndicator();
      },
    );
  }
}

class OtherWidget extends StatelessWidget {
  final List<String>? strings;

  const OtherWidget({Key? key, this.strings}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Text(strings?.first ?? 'Placeholder'); // throws StateError
  }
}
```

**✅ Good:**

```dart
class SomeWidget extends StatelessWidget {
  const SomeWidget({Key? key}) : super(key: key);

  Future<List<String>?> _getStrings() async {
    //... 
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder(
      future: _getStrings(),
      builder: (BuildContext context, AsyncSnapshot<List<String>?> snapshot) {
        if (snapshot.hasData) {
          return OtherWidget(
            strings: snapshot.data?.isNotEmpty == true ? snapshot.data : null,
          );
        }
        return const CircularProgressIndicator();
      },
    );
  }
}

class OtherWidget extends StatelessWidget {
  final List<String>? strings;

  const OtherWidget({Key? key, this.strings}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Text(strings?.first ?? 'Placeholder');
  }
}
```

**Are you willing to submit a pull request to implement this rule?**
Yes

**Comment by incendial** [comment]
@kurtdnyc hi, I'm actually wondering maybe it's better to suggest using `firstOrNull` from `package:collections` https://api.flutter.dev/flutter/package-collection_collection/IterableExtension/firstOrNull.html? Looks like it covers the case and is way easier to implement as a rule. WDYT?

**Comment by kurtdnyc** [comment]
Yeah agreed, that would be much simpler!

**Comment by incendial** [comment]
@kurtdnyc if you need any help with implementing the rule, just let me know!

**Comment by incendial** [comment]
Available in DCM Teams 1.4.0 https://dcm.dev/docs/teams/rules/common/avoid-unsafe-collection-methods/.
